librosa/core/pitch.py

```diff
@@ -333,3 +337,456 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+        time series containing boolean flags indicating whether a frame is voiced or not.
+
+    voiced_prob: np.ndarray [shape=(n_frames,)]
+        time series containing the probability that a frame is voiced.
+
+    See Also
+    --------
+    librosa.yin
+        Fundamental frequency (F0) estimation using the YIN algorithm.
+
+    Examples
+    --------
+    Computing a fundamental frequency (F0) curve from an audio input
+
+    >>> y, sr = librosa.load(librosa.ex('trumpet'))
+    >>> f0, voiced_flag, voiced_probs = librosa.pyin(y, fmin=librosa.note_to_hz('C2'), fmax=librosa.note_to_hz('C7'))
+    >>> times = librosa.times_like(f0)
+
+
+    Overlay F0 over a spectrogram
+
+    >>> import matplotlib.pyplot as plt
+    >>> D = librosa.amplitude_to_db(np.abs(librosa.stft(y)), ref=np.max)
+    >>> fig, ax = plt.subplots()
+    >>> img = librosa.display.specshow(D, x_axis='time', y_axis='log', ax=ax)
+    >>> ax.set(title='pYIN fundamental frequency estimation')
+    >>> fig.colorbar(img, ax=ax, format="%+2.f dB")
+    >>> ax.plot(times, np.where(voiced_flag, f0, np.nan), label='f0', color='w')  # mask unvoiced frames with np.nan
+    >>> ax.legend(loc='upper right')
+    '''
+    # Set the default window length if it is not already specified.
+    if win_length is None:
+        win_length = frame_length // 2
+
+    if win_length > frame_length:
+        raise ParameterError('win_length cannot exceed frame_length')
+
+    # Set the default hop if it is not already specified.
+    if hop_length is None:
+        hop_length = frame_length // 4
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Pad the time series so that frames are centered
+    if center:
+        y = np.pad(y, frame_length // 2, mode=pad_mode)
+
+    # Frame audio.
+    y_frames = util.frame(y, frame_length=frame_length, hop_length=hop_length)
+
+    # Calculate minimum and maximum periods
+    min_period = max(int(np.floor(sr / fmax)), 2)
+    max_period = min(int(np.ceil(sr / fmin)), frame_length-win_length-1)
+
+    # Calculate cumulative mean normalized difference function.
+    yin_frames = _cumulative_mean_normalized_difference(
+        y_frames, frame_length, win_length, min_period, max_period)
+
+    # Parabolic interpolation.
+    parabolic_shifts = _parabolic_interpolation(yin_frames)
+
+    # Find Yin candidates and probabilities.
+    # The implementation here follows the official pYIN software which
+    # differs from the method described in the paper.
+    # 1. Define the prior over the thresholds.
+    thresholds = np.linspace(0, 1, n_thresholds + 1)
+    beta_cdf = scipy.stats.beta.cdf(
+        thresholds, beta_parameters[0], beta_parameters[1])
+    beta_probs = np.diff(beta_cdf)
+
+    yin_probs = np.zeros((yin_frames.shape[0]-2, yin_frames.shape[1]))
+    for i, yin_frame in enumerate(yin_frames.T):
+        # 2. For each frame find the troughs.
+        is_trough = util.localmax(-yin_frame, axis=0)
```

Comment:
Heads up: #1198 adds `util.localmin` so that we can avoid copying each frame here.  I'll let you know when it's merged if you want to rebase and get a bit of a speed boost.

Reply:
ok, localmin is merged now; feel free to use it!

Reply:
Great thanks :+1:. One question I did have: would it be possible to get a version of `localmin` where the first element may be considered a minimum? The reason is that if the actual f0 is slightly above `fmax` then the first element of `yin_frames` would be the best period to pick but it currently isn't marked as a minimum.

Reply:
> would it be possible to get a version of `localmin` where the first element may be considered a minimum?

Nope, this violates the defined semantics.  However, since it returns a mask array, you can always just set the first output like
```python
is_trough[0] = yin_frame[0] <= yin_frame[1]
```